// watcher-daemon/dist/utils/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = void 0;
const dotenv_1 = __importDefault(require("dotenv"));
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
// Load .env file if it exists
const envPath = path_1.default.resolve(process.cwd(), '.env');
if (fs_1.default.existsSync(envPath)) {
    dotenv_1.default.config({ path: envPath });
}
function getEnvVar(key, defaultValue) {
    const value = process.env[key];
    if (value === undefined) {
        if (defaultValue !== undefined) {
            return defaultValue;
        }
        throw new Error(`Missing required environment variable: ${key}`);
    }
    return value;
}
function getEnvBool(key, defaultValue) {
    const value = process.env[key];
    if (value === undefined) {
        return defaultValue;
    }
    return value.toLowerCase() === 'true' || value === '1';
}
function getEnvInt(key, defaultValue) {
    const value = process.env[key];
    if (value === undefined) {
        return defaultValue;
    }
    const parsed = parseInt(value, 10);
    if (isNaN(parsed)) {
        throw new Error(`Environment variable ${key} must be a valid integer`);
    }
    return parsed;
}
exports.config = {
    watchDir: path_1.default.resolve(getEnvVar('WATCH_DIR', './watched')),
    dbPath: path_1.default.resolve(getEnvVar('DB_PATH', './data/rules.db')),
    logFile: path_1.default.resolve(getEnvVar('LOG_FILE', './logs/daemon.log')),
    ollamaHost: getEnvVar('OLLAMA_HOST', 'http://localhost:11434'),
    ollamaModel: getEnvVar('OLLAMA_MODEL', 'llama2'),
    apiEnabled: getEnvBool('API_ENABLED', true),
    apiPort: getEnvInt('API_PORT', 3000),
    notificationsEnabled: getEnvBool('NOTIFICATIONS_ENABLED', true),
    logLevel: getEnvVar('LOG_LEVEL', 'info'),
    watchDebounceMs: getEnvInt('WATCH_DEBOUNCE_MS', 250),
    matchHistoryLimit: getEnvInt('MATCH_HISTORY_LIMIT', 100),
};
exports.default = exports.config;
